rpc.go

```diff
@@ -39,49 +39,76 @@ import (
 
 type (
 	rpcClient struct {
-		ec                 entityConnector
-		client             *amqp.Client
-		clientMu           sync.RWMutex
+		ec     entityConnector
+		client *amqp.Client
+
+		clientMu  sync.RWMutex
+		linkCache map[string]*rpc.Link // stores 'address' => rpc.Link.
+
 		sessionID          *string
 		isSessionFilterSet bool
 		cancelAuthRefresh  func() <-chan struct{}
+
+		// replaceable for testing
+
+		// alias of 'rpc.NewLink'
+		newRPCLink func(conn *amqp.Client, address string, opts ...rpc.LinkOption) (*rpc.Link, error)
+
+		// alias of function 'newAMQPClient'
+		newAMQPClient func(ctx context.Context, ec entityConnector) (*amqp.Client, func() <-chan struct{}, error)
+
+		// alias of 'amqp.Client.Close()'
+		closeAMQPClient func() error
 	}
 
 	rpcClientOption func(*rpcClient) error
 )
 
 func newRPCClient(ctx context.Context, ec entityConnector, opts ...rpcClientOption) (*rpcClient, error) {
 	r := &rpcClient{
-		ec: ec,
+		ec:            ec,
+		linkCache:     map[string]*rpc.Link{},
+		newRPCLink:    rpc.NewLink,
+		newAMQPClient: newAMQPClient,
 	}
 
+	r.closeAMQPClient = func() error { return r.client.Close() }
```

Comment:
You should be able to shorten this to `r.closeAMQPClient = r.client.Close`

Reply:
🤦 

Yes.

Reply:
I remember why I did this now.

r.client isn't defined at this point (ie, this needs to be lazy evaluated to grab the latest value of `r.client`)

Reply:
Ah ok makes sense.